### ISSUE TITLE
[Feature Request] User permissions based on custom column (virtual libraries?)

### ISSUE THREAD

**Issue by Katzah** [janeczku/calibre-web#800]
I would like to be able to white-list values in a custom column to give users permission to only certain things.

The mature content setting is good for the purpose it actually is intended for but not flexible enough for my use as I need it to be an actual mature content setting. I'm running two libraries and two calibe-web instances at the moment in order to only make part of the library visible to most users but I'm obviously duplicating a lot of files, work and processes here that don't really need to be.

If you could create blacklist or white-list permissions based on both custom columns and tags that would be ideal but definitely the custom columns if this is possible please.

**Comment by OzzieIsaacs** [comment]
Releated to #547

**Comment by OzzieIsaacs** [comment]
I don't understand the problem excacly and you proposed solution, could you give an example what is not not working. 
What would be better with a custom column compared to tags? You can have more than one tag attached to a book. So one tag could act as marker for visibility and other tags could be used as "normal" tags. 
If a custom column would be used, should it be a numeric column or comma seperated texts like now?
Currently I plan to implement is like described in #547. On user base you can add tags which are not displayed ("blacklist" only)

**Comment by Katzah** [comment]
What you're suggesting would be ok if the only use for permissions was to block a specific genre or similar from people who shouldn't see it which is more useful for the content filter like you already have or something simple like some people don't want non-fiction. It doesn't really lend itself to customised libraries for users or groups of users. As an example I have several users who only want to see certain things. I'm a little over 30 and my grandparents are a couple of my users. They are obviously a bit older and not really tech savvy so just throwing the lot at them and telling them to use the search functions to find what they want is kind of overwhelming. (The fact that they have a tablet and e-readers is a giant leap into the future, let's not scare them off right away.)

This whole principle also relies on the tag being present to exclude anything but tagging is inherently manual and personal. That makes it extremely prone to user error and inconsistency, simple spelling mistakes, forgetting to add something, using a different word, etc. Although I'm sure most of us try to be consistent, tags are inherently meant to be descriptive not necessarily specific. If you're trying to block more than one or two things then it's a lot of tags, a very manual process and a pain in the ass to implement on an ongoing, day to day basis. (Every time you add a book you have to think through all your different users and their permissions.)

For me as the admin, reliably blacklisting 99% of a library with thousands of books in it for just one person because they only want 3 authors becomes a nightmare because you have to remember to add that tag to every book that they don't want. If you then have another reader that only wants a certain genre and another that's only following a few series than blacklisting everything except what they want for each person is really not viable. I'd also have to implement a #notanyone automatic tag for new books on everyone except me if I didn't want every new book coming in from Lazylib to automatically go to everyone. This would be annoying to implement because of the way I'm bringing books in and I'd still be faced with a messy tagging situation so it's not going to work for me at all. 

However if you can apply some kind of operation like =, <>, true/false, particular option from a list, text contains/text does not contain to a custom column you can effectively implement virtual libraries from the system admin point of view. One of two of these would be ample logic control when combined with the full capability of creating custom columns in calibre.

What I would like to be able to do is set up one or more virtual libraries for each person within calibre itself and put the virtual library names in a column. Set one permission for each user in calibre-web (virtual library column contains their user name) and any book that fits their search or searches goes to them automatically as soon as I mark it public. 

Basically... columns would be exponentially more powerful and easier to work with from the system admin end. Instead of having to add a ton of tags to every book or block everything for everyone and then unblock each person on each book they want I can just check the book and metadata, mark it public and it'll go to all the people who want it based on their searches.

**Comment by Katzah** [comment]
I was thinking this over and it occurs to me that the better option goes a step further than this. Essentially what I'm after is permissions based on virtual libraries but I know from reading elsewhere you can't fully implement this in calibre-web due to the unique search language in calibre. However while this means you can't implement virtual libraries from a user perspective you can use this concept to implement virtual libraries as set up by the library admin in calibre which is something a lot of people have asked for. It could work something like this:

1. in your mains settings have an option to enable the calibre virtual libraries in calibre-web
Either you create a custom column in calibre with the correct template and name that you're expecting or if this isn't possible you put a note there along the lines of "custom column in calibre listing applicable virtual libraries is required for this to work", link it to your wiki with a quick description of how to do that

2. create a list of virtual libaries
You might be able to fetch these from calibre directly? If not you can build it from the values in the column.

3. a book matches a virtual libary if it has the name of the library in the custom column

Once you have that structure there I can think of two things off the bat that would be amazing. First is in the user settings the option to give user access to all books or restrict to selected libraries. Second is in the book browser, a big drop down on the top left where you can select all books or any of the virtual libraries you have access to. This would make browsing large libraries much more friendly.

The setup of virtual libraries would need to be done in calibre itself but you could build whatever functionality you wanted over the top of it in calibre-web.
